### ISSUE TITLE
Declaring memoize val in Spek root causes it to produce different values on given/when stages

### ISSUE THREAD

**Issue by Lingviston** [JetBrains/spek#652]
I've just started investigating Spek framework, so please forgive me inconvenient usage of some terms.

Following Gherkin style I want to have several `Feature`s within a single Spek test. I want all mocks and OUT (object under test) instance to be shared between all of the features so I need to declare them on the root Spek object level.

Looking at `SetSpek `example [here](https://spekframework.org/migration/) I assume that it is possible to put memoized vars on the root Spek level.

But when I started to implement tests this way I found our that my memoized vars are not the same instances in `Given` and `When` lambdas. 

Here's a code snippet, which demonstrates the problem:

```
object HoldDatesRepositoryImplTest : Spek({

    val calendarFactory by memoized { mock<CalendarFactory>() }

    Feature("Default end time") {

        Scenario("Start time < 7 PM") {

            Given("Start time is 4:05 PM") {
                System.out.println("Given $calendarFactory")
            }

            When("Subscribed to dates") {
                System.out.println("When $calendarFactory")
            }
        }
    }
}
```
The outputs are:

> Given Mock for CalendarFactory, hashCode: 836427078
> When Mock for CalendarFactory, hashCode: 834466358

However if I declare memoized var within `Feature` it works as expected, i.e. all calls to the var produce the same value. Here's code snippet:
```
object HoldDatesRepositoryImplTest : Spek({

    Feature("Default end time") {

        val calendarFactory by memoized { mock<CalendarFactory>() }

        Scenario("Start time < 7 PM") {

            Given("Start time is 4:05 PM") {
                System.out.println("Given $calendarFactory")
            }

            When("Subscribed to dates") {
                System.out.println("When $calendarFactory")
            }
        }
    }
}
```

The outputs are:

> Given Mock for CalendarFactory, hashCode: 1016860054
> When Mock for CalendarFactory, hashCode: 1016860054

So, my questions are:

1. Are several `Features` within the same `Spek` supported? I checked that they are, but then I got the second and the most important question.
2. How to share mocks and OUT instances between several `Feature`s in the same `Spek`?

**Comment by raniejade** [comment]
> Are several Features within the same Spek supported? I checked that they are, but then I got the second and the most important question.

Yes

> How to share mocks and OUT instances between several Features in the same Spek?

Outside `Feature` the default caching mode is `CachingMode.TEST` which means all `test` scope will get a unique instance (Given, When, etc.. are test scopes). Within a `Feature` the default is `CachingMode.GROUP` which means group scopes (`Scenario` is a group scope) will have a unique instance. Depending on what you want you can use `CachingMode.GROUP` means all `Scenario`s receives a unique instance or `CachingMode.SCOPE` means all your `Feature` will receive a unique instance (a `Scenario` within a `Feature` shares the same instance). Ideally, treat each `Feature` separately and don't share anything.

**Comment by Lingviston** [comment]
@raniejade thank you for claryfying the group vs. test scope for me. Reading the [doc](https://spekframework.org/core-concepts/) I could not properly map them to the Gherkin style:(

> Ideally, treat each Feature separately and don't share anything.
That's what I ended up with. 

`CachingMode.SCOPE` is not what I want as it leads to the same object being shared across different features. What I need is unique object per scenario, but declared outside of the feature. I checked and looks like declaration with `CachingMode.GROUP` does what I need.

Tested with the following setup:
Spek
|- Feature
|--- Scenario
|--- Scenario
|- Feature
|--- Scenario
|--- Scenario

Tests within each scenario were sharing instance unique per scenario. I wonder, wouldn't it be more convenient to make `CachingMode.GROUP` default outside of the `Feature`?

**UPDATE** Is there a way to override/define default caching mode for the scope?

**Comment by raniejade** [comment]
> Tests within each scenario were sharing instance unique per scenario. I wonder, wouldn't it be more convenient to make CachingMode.GROUP default outside of the Feature?

`CachingMode.TEST` is the more practical default as a good practice tests should not share anything.

> Is there a way to override/define default caching mode for the scope?

No, if you're not using the core dsl (`group`, `test`, ...). And the core DSL is not intended to be use directly - it is there to provide infrastructure to build styles (specification, gherkin, ...)

**Comment by raniejade** [comment]
I'll re-purpose this ticket to improve the documentation about `CachingMode` and `memoized`.

**Comment by Lingviston** [comment]
> CachingMode.TEST is the more practical default as a good practice tests should not share anything.

But since Given/When/Then are actually test scopes - isn't that incorrect? I mean they're tests but they need to share the data. At least because in `Given` we usually set expectations for the mocks and in `Then` we verify them on the same object.

Prior to Spek those tests could be written with JUnit4 in the following manner:

```
class SomeTest {
    val someMock = mock()

    val tested = SomeTestedClass(someMock)

    @Test
    fun `given smth when smth then smth`() {
    }

    @Test
    fun `given smth else when smth then smth else happens`() {
    }
}
```
So as you can see all mocks and OUT are declared one time and links could be shared between tests, but thanks to the JUnit4 (it creates new instance of test suite class to execute each test) all tests get new instances of mocks and OUT.

All I wanted to achieve was mapping this structure to the Spek. I think I did it correct and only introduced `Feature`s for better grouping of my test scenarios. But after this direct mapping it didn't work because in Spek `Given`/`When`/`Then`are tests and `memoized` uses `Test` `CacheMode` by default. 

That's why I think having Group mode for root Spek by default is more straightforward at least for those, who migrate from JUnit 4 as there Scenarios were actually test scopes.

**Comment by raniejade** [comment]
> But since Given/When/Then are actually test scopes - isn't that incorrect? I mean they're tests but they need to share the data. At least because in Given we usually set expectations for the mocks and in Then we verify them on the same object.

They share data/state if you declare the `memoized` within `Feature` and it is recommended way.

Your example exactly maps to:
```kotlin
Feature("SomeTest") {
  val someMock by memoized { ... }
  val tested by memoized { SomeTestedClass(someMock) }

  Scenario("given smth when smth then smth") { ... }

  Scenario("given smth else when smth then smth else happens") { ... }
}
```

Also spek supports another style: `specification` which the root default makes sense (`CachingMode.TEST`). Say you're a weird person that likes to use the two styles in a single class:

```kotlin
val someMock by memoized { ... }
Feature("...") { ... }
describe("...") { ... }
```

`CachingMode.GROUP` makes sense for `Feature` but not for `describe` and vice versa. There's no  correct default that would satisfy both styles, so I had to choose one and just recommend not declaring `memoized` on the root scope (not explicit but most of the examples does not declare `memoized` outside of the styles).

**Comment by Lingviston** [comment]
Ok, in my example I used the same `When` part. But it can be different as well and thus lead to different `Features`. 

Here's a real life example of what I covered with Spek tests. I have a publisher class, which gives you Reactive stream of date updates. This stream has initial value, which is dependent on the current date. Publisher also allows you to set new date value, which is processed somehow, and then emitted by the stream. So technically I have two features: default date, change date. In both features I have two scenarios: 
1) Default date: if current time is less than 7 PM or not;
2) Setting new date: if new date set is less than current one or not;

So I have 4 scenarios, grouped into 2 features. And all those scenarios need the same set of mocks and OUT. They just need to set different expectations from mocks, but they don't need to declare them twice. With the JUnit4 approach that I demonstrated above I would just add 2 more tests into the same class and they could share all of the setup. With the Spek I need to either duplicate mocks/OUT declaration in Features or define them in root scope but explicitly change the caching mode.

I understand that this is how Spek works and that I have to deal with it, but I just want to share my experience during migration to Spek and how it might be confusing to the person, who was only working with JUnit4 tests before. At the same time confusion could be gone if default CachingMode was Group. It might not fit Specification style as you said, but from my POV it better fits Gherkin. So, perhaps Spek framework can be improved in this area.
